qi-lib/flow/core/pass.rkt

```diff
@@ -0,0 +1,71 @@
+#lang racket/base
+
+(provide find-and-map/qi
+         fix)
+
+(require racket/match
+         syntax/parse
+         "private/form-property.rkt")
+
+;; Utilities that are used in each compiler pass
+
+;; Walk the syntax tree in a "top down" manner, i.e. from the root down
+;; to the leaves, applying a transformation to each node.
+;; The transforming function is expected to either return transformed
+;; syntax or false.
+;; The traversal terminates at a node if either the transforming function
+;; "succeeds," returning syntax different from the original, or if it
+;; returns false, indicating that the node should not be explored.
+;; In the latter case, the node is left unchanged.
+;; Otherwise, as long as the transformation is the identity, it will continue
+;; traversing subexpressions of the node.
+(define (find-and-map f stx)
+  ;; f : syntax? -> (or/c syntax? #f)
+  (match stx
+    [(? syntax?) (let ([stx^ (f stx)])
+                   (if stx^
+                       (if (eq? stx^ stx)
+                           ;; no transformation was applied, so
+                           ;; keep traversing
+                           (datum->syntax stx
+                             (find-and-map f (syntax-e stx))
+                             stx
+                             stx)
+                           ;; transformation was applied, so we stop
+                           stx^)
+                       ;; false was returned, so we stop
+                       stx))]
+    [(cons a d) (cons (find-and-map f a)
+                      (find-and-map f d))]
+    [_ stx]))
+
+;; A thin wrapper around find-and-map that does not traverse subexpressions
+;; that are tagged as host language (rather than Qi) expressions
+(define (find-and-map/qi f stx)
+  ;; #%host-expression is a Racket macro defined by syntax-spec
+  ;; that resumes expansion of its sub-expression with an
+  ;; expander environment containing the original surface bindings
+  ;; TODO: technically should be ~literal host expression to not
+  ;; collide with a user-defined #%host-expression binding, but that
+  ;; would never be hit in practice since that would be rewritten
+  ;; through expansion to a use of the core language. In general,
+  ;; we should be using ~literal matching throughout the compiler.
```

Comment:
Now that we rely on the syntax property to only apply transformations at proper form positions, I don't think we need to use `~literal` matching anymore. I think the use of `~literal` matching was an early attempt to avoid misinterpreting binding positions before we realized that other overlaps could happen like the `(partition [tee ...])` case and thus that we needed a different strategy.